File: web/packages/loom-texture/src/dataset/export.ts
import { canonicalBytes, canonicalJson } from '../canonical-json.js';
import { LIBRARY, definitionOf } from '../catalog.js';
import { sha256Hex } from '../digest.js';
import { stream } from '../hash.js';
import { floorDiv } from '../integer.js';
import { LICENCE_ID } from '../licence.js';
import { bakeMaps } from '../maps.js';
import { BAKE_PIPELINE } from '../objects.js';
import type { Recipe } from '../recipe.js';
import { DATASET_PROFILE, type DatasetPlan, RENDERER, SAMPLER } from './plan.js';
import { type Light, renderView, sampleLight, sampleView } from './render.js';
import { sampleRecipe } from './sample.js';

/**
 * Export a dataset plan: image shards, one line per record, and the manifest that pins them.
 *
 * The directory holds `images/<set id>.rgb` (every record of that set, `image_size` squared
 * pixels of sRGB each, concatenated in record order), `records.jsonl` (one canonical JSON object
 * per line, in record order), and `dataset.json`, the manifest: the plan, the generator versions,
 * the makers and published recipes the records came from, and the length and sha256 of every
 * file. The manifest is written last and is the one file a repository keeps; the bytes stay out
 * of git, and anyone can rebuild them from the plan and check them against it.
 *
 * Records are numbered across the whole dataset, and each record's seed is a stream of the plan's
 * seed and its number, so a record is the same whichever machine, runtime or shard order made it.
 */
export const IMAGE_DIRECTORY = 'images';
export const RECORDS_FILE = 'records.jsonl';
export const DATASET_FILE = 'dataset.json';

export interface DatasetRecord {
  readonly index: number;
  readonly set_id: string;
  /** Position of this record's picture within its set's shard. */
  readonly offset: number;
  readonly recipe: Recipe;
  readonly recipe_sha256: string;
  readonly view: {
    readonly u: number;
    readonly v: number;
    /** Surface under one texel, in micrometres, so a picture's physical scale is known. */
    readonly texel_um: { readonly u: number; readonly v: number };
  };
  readonly light: Light;
  /** Attempts the sampler used, the accepted one included. */
  readonly attempts: number;
}

export interface ShardEntry {
  readonly set_id: string;
  readonly path: string;
  readonly records: number;
  readonly byte_size: number;
  readonly sha256: string;
}

/** Receives each file once, complete, in the order the export finishes them. */
export type Sink = (path: string, bytes: Uint8Array) => void;

export interface DatasetExport {
  readonly manifest: Uint8Array;
  readonly records: number;
  readonly refused: Readonly<Record<string, number>>;
}

/** The seed a record draws everything from: a stream of the plan's seed and the record's number. */
export const recordSeed = (plan: DatasetPlan, index: number): number => stream(plan.seed, index);

export function exportDataset(plan: DatasetPlan, sink: Sink): DatasetExport {
  const pixels = plan.image_size * plan.image_size * 3;
  const lines: string[] = [];
  const shards: ShardEntry[] = [];
  const refused: Record<string, number> = {};
  const makers = new Map<string, { maker_id: string; version: number; object_sha256: string }>();
  const bases: { set_id: string; version: number; recipe_sha256: string }[] = [];
  let index = 0;
  for (const setId of plan.sets) {
    const source = LIBRARY.find((candidate) => candidate.entry.set_id === setId);
    if (source === undefined) throw new Error(`${setId} is not a published set`);
    const { manifest } = source.maker;
    makers.set(`${manifest.maker_id} ${manifest.version}`, {
      maker_id: manifest.maker_id,
      version: manifest.version,
      object_sha256: sha256Hex(canonicalBytes(manifest)),
    });
    bases.push({
      set_id: setId,
      version: source.entry.version,
      recipe_sha256: sha256Hex(canonicalBytes(source.entry.recipe)),
    });
    const images = new Uint8Array(plan.records_per_set * pixels);
    let retries = 0;
    for (let offset = 0; offset < plan.records_per_set; offset += 1, index += 1) {
      const seed = recordSeed(plan, index);
      const { recipe, attempts } = sampleRecipe(plan, source, seed);
      retries += attempts - 1;
      const maps = bakeMaps(definitionOf({ ...source, entry: { ...source.entry, recipe } }));
      const view = sampleView(seed, maps.width, maps.height);
      const light = sampleLight(plan, seed);
      images.set(renderView(maps, view, light, plan.image_size), offset * pixels);
      const record: DatasetRecord = {
        index,
        set_id: setId,
        offset,
        recipe,
        recipe_sha256: sha256Hex(canonicalBytes(recipe)),
        view: {
          u: view.u,
          v: view.v,
          texel_um: {
            u: floorDiv(recipe.extent_mm.u * 1000, maps.width),
            v: floorDiv(recipe.extent_mm.v * 1000, maps.height),
          },
        },
        light,
        attempts,
      };
      lines.push(canonicalJson(record));
    }
    const path = `${IMAGE_DIRECTORY}/${setId}.rgb`;
    sink(path, images);
    shards.push({
      set_id: setId,
      path,
      records: plan.records_per_set,
      byte_size: images.length,
      sha256: sha256Hex(images),
    });
    refused[setId] = retries;
  }
  const records = new TextEncoder().encode(lines.map((line) => `${line}\n`).join(''));
  sink(RECORDS_FILE, records);
  const manifest = canonicalBytes({
    profile: DATASET_PROFILE,
    truth: 'invented',
    licence_id: LICENCE_ID,
    plan,
    plan_sha256: sha256Hex(canonicalBytes(plan)),
    generator: { pipeline: BAKE_PIPELINE, sampler: SAMPLER, renderer: RENDERER },
    makers: [...makers.values()].sort((a, b) =>
      a.maker_id < b.maker_id ? -1 : a.maker_id > b.maker_id ? 1 : a.version - b.version,
    ),
    bases,
    image: {
      width: plan.image_size,
      height: plan.image_size,
      channels: 3,
      layout: 'sRGB bytes, rows top to bottom, one picture after another in record order',
    },
    shards,
    records: {
      path: RECORDS_FILE,
      count: index,
      byte_size: records.length,
      sha256: sha256Hex(records),
    },
    refused_attempts: refused,
  });
  sink(DATASET_FILE, manifest);
  return { manifest, records: index, refused };
}

File: ml/exulanica_training/texture_inverse/export.py
"""A synthetic texture dataset export, read back and verified before a single picture is used.

``web/packages/loom-texture`` exports a plan into ``images/<set id>.rgb`` shards,
``records.jsonl`` and ``dataset.json``, and the repository commits only ``dataset.json``
(``web/packages/loom-texture/dataset/manifests/``). A training run is given the export and the
committed manifest, and refuses unless:

- the export's ``dataset.json`` is the committed manifest, byte for byte;
- the manifest says ``"truth": "invented"`` and ``"licence_id": "CC0-1.0"``, so nothing observed
  and nothing private to a workspace is ever trained on by this path;
- every shard and the records file have the length and sha256 the manifest names;
- every record is canonical JSON, numbered in order, and names a shard offset that exists.

This module is plain Python: it runs, and its tests run, without torch.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Final

__all__ = [
    "DATASET_FILE",
    "DATASET_PROFILE",
    "ExportRefused",
    "Record",
    "VerifiedExport",
    "canonical_bytes",
    "verify_export",
]

DATASET_PROFILE: Final = "exulanica.texture-dataset/v1"
DATASET_FILE: Final = "dataset.json"
RECORDS_FILE: Final = "records.jsonl"
#: The only licence this path trains on. A workspace's private bakes are never a training input.
TRAINABLE_LICENCE: Final = "CC0-1.0"
_SHA256_HEX: Final = frozenset("0123456789abcdef")


class ExportRefused(RuntimeError):
    """The export is not the dataset its committed manifest describes."""


def canonical_bytes(value: Any) -> bytes:
    """The repository's canonical JSON: sorted keys, no spaces, integers only."""

    def check(node: Any, path: str) -> None:
        if isinstance(node, float):
            raise ExportRefused(f"{path} is a float, which no dataset object holds")
        if isinstance(node, Mapping):
            for key, item in node.items():
                check(item, f"{path}.{key}")
        elif isinstance(node, list):
            for index, item in enumerate(node):
                check(item, f"{path}[{index}]")

    check(value, "$")
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=True).encode(
        "ascii"
    )


def _sha256(raw: bytes) -> str:
    return hashlib.sha256(raw).hexdigest()


def _is_sha256(value: object) -> bool:
    return isinstance(value, str) and len(value) == 64 and set(value) <= _SHA256_HEX


def _strict(raw: bytes, where: str) -> Any:
    def pairs(items: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, item in items:
            if key in result:
                raise ExportRefused(f"{where} repeats the key {key!r}")
            result[key] = item
        return result

    def refuse_float(literal: str) -> Any:
        raise ExportRefused(f"{where} holds the non-integer {literal}")

    try:
        document = json.loads(
            raw.decode("utf-8"), object_pairs_hook=pairs, parse_float=refuse_float
        )
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ExportRefused(f"{where} is not JSON") from error
    if canonical_bytes(document) != raw:
        raise ExportRefused(f"{where} is not canonical JSON")
    return document


@dataclass(frozen=True, slots=True)
class Record:
    """One training pair: where its picture is, and the recipe the picture was baked from."""

    index: int
    set_id: str
    offset: int
    recipe: Mapping[str, Any]
    recipe_sha256: str


@dataclass(frozen=True, slots=True)
class VerifiedExport:
    """An export whose every file matched its committed manifest."""

    directory: Path
    manifest: Mapping[str, Any]
    manifest_sha256: str
    image_size: int
    records: tuple[Record, ...]
    #: Shard path by set id, relative to the directory.
    shards: Mapping[str, str]

    def maker_objects(self) -> tuple[tuple[str, int, str], ...]:
        """``(maker id, version, manifest object sha256)`` for every maker the records use."""
        return tuple(
            (maker["maker_id"], maker["version"], maker["object_sha256"])
            for maker in self.manifest["makers"]
        )

    def picture(self, record: Record) -> bytes:
        """The record's sRGB bytes, rows top to bottom, read from its shard."""
        pixels = self.image_size * self.image_size * 3
        with (self.directory / self.shards[record.set_id]).open("rb") as handle:
            handle.seek(record.offset * pixels)
            data = handle.read(pixels)
        if len(data) != pixels:
            raise ExportRefused(f"record {record.index} runs past its shard")
        return data


def _file(directory: Path, relative: str, size: int, digest: str) -> None:
    path = directory / relative
    if path.is_symlink() or not path.is_file():
        raise ExportRefused(f"{relative} is not a regular file in the export")
    if path.stat().st_size != size:
        raise ExportRefused(f"{relative} is {path.stat().st_size} bytes, not {size}")
    hasher = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1 << 20), b""):
            hasher.update(chunk)
    if hasher.hexdigest() != digest:
        raise ExportRefused(f"{relative} does not hash to its manifest's digest")


def verify_export(directory: Path, committed_manifest: Path) -> VerifiedExport:
    """The export at ``directory``, checked file by file against ``committed_manifest``."""
    committed = committed_manifest.read_bytes()
    exported = (directory / DATASET_FILE).read_bytes()
    if exported != committed:
        raise ExportRefused(f"{DATASET_FILE} is not the committed manifest, byte for byte")
    manifest = _strict(committed, DATASET_FILE)
    if manifest.get("profile") != DATASET_PROFILE:
        raise ExportRefused(f"the manifest is not {DATASET_PROFILE}")
    if manifest.get("truth") != "invented":
        raise ExportRefused("only invented pictures are trained on by this path")
    if manifest.get("licence_id") != TRAINABLE_LICENCE:
        raise ExportRefused(f"only {TRAINABLE_LICENCE} data is trained on by this path")
    image = manifest["image"]
    if image.get("channels") != 3 or image.get("width") != image.get("height"):
        raise ExportRefused("pictures are square and three-channel")
    size = image["width"]
    shards: dict[str, str] = {}
    counts: dict[str, int] = {}
    for shard in manifest["shards"]:
        if not _is_sha256(shard["sha256"]) or shard["path"] != f"images/{shard['set_id']}.rgb":
            raise ExportRefused(f"shard {shard['set_id']} is not named as the exporter names it")
        if shard["byte_size"] != shard["records"] * size * size * 3:
            raise ExportRefused(f"shard {shard['set_id']} is not whole pictures")
        _file(directory, shard["path"], shard["byte_size"], shard["sha256"])
        shards[shard["set_id"]] = shard["path"]
        counts[shard["set_id"]] = shard["records"]
    listing = manifest["records"]
    _file(directory, listing["path"], listing["byte_size"], listing["sha256"])
    lines = (directory / listing["path"]).read_bytes().split(b"\n")
    if lines[-1] != b"" or len(lines) - 1 != listing["count"]:
        raise ExportRefused("the records file does not hold the counted records, one per line")
    records: list[Record] = []
    for index, line in enumerate(lines[:-1]):
        document = _strict(line, f"record {index}")
        recipe = document["recipe"]
        if (
            document["index"] != index
            or document["set_id"] not in shards
            or not 0 <= document["offset"] < counts[document["set_id"]]
            or _sha256(canonical_bytes(recipe)) != document["recipe_sha256"]
        ):
            raise ExportRefused(f"record {index} does not agree with its position or its recipe")
        records.append(
            Record(
                index=index,
                set_id=document["set_id"],
                offset=document["offset"],
                recipe=recipe,
                recipe_sha256=document["recipe_sha256"],
            )
        )
    return VerifiedExport(
        directory=directory,
        manifest=manifest,
        manifest_sha256=_sha256(committed),
        image_size=size,
        records=tuple(records),
        shards=shards,
    )


def iter_split(export: VerifiedExport, *, held_out_every: int) -> Iterator[tuple[str, Record]]:
    """Each record with ``train`` or ``held_out``, decided by its number and nothing else."""
    for record in export.records:
        yield ("held_out" if record.index % held_out_every == 0 else "train"), record
